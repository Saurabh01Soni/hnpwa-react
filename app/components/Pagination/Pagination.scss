$_pagination-box-shadow-color: rgba(0, 0, 0, .1);

.Pagination {
  box-shadow: 0 1px 2px $_pagination-box-shadow-color;
  height: $header-height;
  text-align: center;
  width: 100%;
}

.Pagination__inner {
  display: inline-block;
  padding: 18px 0;
}

.Pagination__button {
  background-color: transparent;
  border: 0;
}

.Pagination__button--disabled {
  color: $color-gray;
}

.Pagination__number {
  display: inline-block;
  margin: 0 20px;
}
