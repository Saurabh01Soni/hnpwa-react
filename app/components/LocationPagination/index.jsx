import React from 'react';
import PropTypes from 'prop-types';

import Pagination from 'components/Pagination';

const propTypes = {
  currentPage: PropTypes.number,
  location: PropTypes.shape({ pathname: PropTypes.string }).isRequired,
  onPaginate: PropTypes.func,
};

const defaultProps = {
  currentPage: 1,
  onPaginate() {},
};

class LocationPagination extends React.Component {
  static getRequestQuery(pathname) {
    return pathname.split('/').filter(v => !!v);
  }

  constructor(props) {
    super(props);
    this.handlePaginate = this.handlePaginate.bind(this);
  }

  componentWillMount() {
    const [type, stringPage] = LocationPagination.getRequestQuery(this.props.location.pathname);
    const page = parseInt(stringPage, 10);

    if (type && !Number.isNaN(page)) {
      this.props.onPaginate(type, parseInt(page, 10));
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.location.pathname !== this.props.location.pathname) {
      const [type, page] = LocationPagination.getRequestQuery(nextProps.location.pathname);
      this.props.onPaginate(type, parseInt(page, 10));
    }
  }

  handlePaginate(page) {
    const [type] = LocationPagination.getRequestQuery(this.props.location.pathname);
    this.props.onPaginate(type, page);
  }

  render() {
    const { currentPage } = this.props;

    return (
      <Pagination
        curentPage={currentPage}
        onPaginate={this.handlePaginate}
      />
    );
  }
}

LocationPagination.propTypes = propTypes;
LocationPagination.defaultProps = defaultProps;

export default LocationPagination;
