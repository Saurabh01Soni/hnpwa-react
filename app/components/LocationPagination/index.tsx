import * as React from 'react';

import Pagination from 'components/Pagination';

interface ILocationPaginationProps {
  currentPage?: number;
  done?: () => void;
  feedCount?: number;
  history: any;
  location: any;
  onPaginate?: (type: string, page: number) => Promise<any>;
}

class LocationPagination extends React.Component<ILocationPaginationProps> {
  public static defaultProps = {
    currentPage: 1,
    done() {},
    feedCount: 0,
    history: {},
    location: {},
    onPaginate: (type: string, page: number) => Promise.resolve(),
  };

  public static getRequestQuery(pathname: string): string[] {
    return pathname.split('/').filter(v => !!v);
  }

  public constructor(props: ILocationPaginationProps) {
    super(props);
    this.handlePaginate = this.handlePaginate.bind(this);
  }

  public componentWillMount() {
    const { done, feedCount } = this.props;
    const [type, stringPage] = LocationPagination.getRequestQuery(this.props.location.pathname);
    const page = parseInt(stringPage, 10);

    if (type && !Number.isNaN(page) && feedCount === 0) {
      this.props.onPaginate(type, page).then(done, done);
    }
  }

  public componentWillReceiveProps(nextProps: ILocationPaginationProps) {
    if (nextProps.location.pathname !== this.props.location.pathname) {
      const [type, page] = LocationPagination.getRequestQuery(nextProps.location.pathname);
      this.props.onPaginate(type, parseInt(page, 10));
    }
  }

  private handlePaginate(page: number) {
    const [type] = LocationPagination.getRequestQuery(this.props.location.pathname);
    this.props.history.push(`/${type}/${page}`);
  }

  public render() {
    const { currentPage } = this.props;

    return (
      <Pagination
        currentPage={currentPage}
        onPaginate={this.handlePaginate}
      />
    );
  }
}

export default LocationPagination;
