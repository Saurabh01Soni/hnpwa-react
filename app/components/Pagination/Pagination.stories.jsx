import React from 'react';
import { storiesOf } from '@storybook/react';

import Pagination from './';

const stories = storiesOf('Pagination', module);

class PaginationWrapper extends React.Component {
  constructor(props) {
    super(props);
    this.state = { currentPage: 1, maxPage: 10 };
    this.handlePaginate = this.handlePaginate.bind(this);
  }

  handlePaginate(number) {
    this.setState({ currentPage: number });
  }

  render() {
    return (
      <Pagination
        currentPage={this.state.currentPage}
        maxPage={this.state.maxPage}
        onPaginate={this.handlePaginate}
      />
    );
  }
}

stories
  .add('Pagination', () => <PaginationWrapper />);
