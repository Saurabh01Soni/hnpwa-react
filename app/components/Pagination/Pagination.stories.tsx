import * as React from 'react';
import { storiesOf } from '@storybook/react';

import Pagination from './';

const stories = storiesOf('Pagination', module);

interface IPaginationWrapperState {
  currentPage: number;
}

class PaginationWrapper extends React.Component<{}, IPaginationWrapperState> {
  constructor(props) {
    super(props);
    this.state = { currentPage: 1 };
    this.handlePaginate = this.handlePaginate.bind(this);
  }

  public handlePaginate(value: number) {
    this.setState({ currentPage: value });
  }

  public render() {
    return (
      <Pagination
        currentPage={this.state.currentPage}
        onPaginate={this.handlePaginate}
      />
    );
  }
}

stories
  .add('Pagination', () => <PaginationWrapper />);
