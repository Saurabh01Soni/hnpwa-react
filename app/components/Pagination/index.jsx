import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

import './Pagination.scss';

const propTypes = {
  currentPage: PropTypes.number,
  maxPage: PropTypes.number,
  onPaginate: PropTypes.func,
};

const defaultProps = {
  currentPage: 1,
  maxPage: 1,
  onPaginate() {},
};

function Pagination({ currentPage, maxPage, onPaginate }) {
  const prevDisabled = currentPage === 1;
  const nextDisabeld = currentPage === maxPage;

  const prevButtonClassName = cx('Pagination__button', {
    'Pagination__button--disabled': prevDisabled,
  });

  const nextButtonClassName = cx('Pagination__button', {
    'Pagination__button--disabled': nextDisabeld,
  });

  return (
    <div className="Pagination">
      <div className="Pagination__inner">
        <button
          className={prevButtonClassName}
          disabled={prevDisabled}
          type="button"
          onClick={() => onPaginate(currentPage - 1)}
        >
          {'<'} Prev
        </button>
        <div className="Pagination__number">{currentPage} / {maxPage}</div>
        <button
          disabled={nextDisabeld}
          className={nextButtonClassName}
          type="button"
          onClick={() => onPaginate(currentPage + 1)}
        >
          {'>'} Next
        </button>
      </div>
    </div>
  );
}

Pagination.propTypes = propTypes;
Pagination.defaultProps = defaultProps;

export default Pagination;
