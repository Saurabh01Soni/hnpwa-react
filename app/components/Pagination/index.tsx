import * as cx from 'classnames';
import * as React from 'react';

import './Pagination.scss';

export interface IPaginationProps {
  readonly currentPage?: number;
  readonly onPaginate?: (value: number) => void;
}

const defaultProps: IPaginationProps = {
  currentPage: 1,
  onPaginate() {},
};

const Pagination: React.SFC<IPaginationProps> = ({ currentPage, onPaginate }) => {
  const prevDisabled = currentPage === 1;
  const nextDisabled = currentPage === 10;

  const prevButtonClassName = cx('Pagination__button', {
    'Pagination__button--disabled': prevDisabled,
  });

  const nextButtonClassName = cx('Pagination__button', {
    'Pagination__button--disabled': nextDisabled,
  });

  return (
    <div className="Pagination">
      <div className="Pagination__inner">
        <button
          className={prevButtonClassName}
          data-button="prev"
          disabled={prevDisabled}
          type="button"
          onClick={() => onPaginate(currentPage - 1)}
        >
          {'<'} Prev
        </button>
        {currentPage} / 10
        {' '}
        <button
          className={nextButtonClassName}
          data-button="next"
          disabled={nextDisabled}
          type="button"
          onClick={() => onPaginate(currentPage + 1)}
        >
          Next {'>'}
        </button>
      </div>
    </div>
  );
};

Pagination.defaultProps = defaultProps;

export default Pagination;
