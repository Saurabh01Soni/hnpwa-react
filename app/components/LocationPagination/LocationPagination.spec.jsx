import React from 'react';
import { shallow } from 'enzyme';

import Pagination from 'components/Pagination';
import LocationPagination from './';

const history = { push() {} };
const location = { pathname: '/news/1' };

describe('<LocationPagination />', () => {
  it('should match snapshot when render default', () => {
    const wrapper = shallow(<LocationPagination history={history} location={location} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should return query array when calls LocationPagination.getRequestQuery', () => {
    expect(LocationPagination.getRequestQuery('/news/1')).toEqual(['news', '1']);
  });

  it('should calls props.onPaginate in componentWillMount when URL path is valid and props.feedCount === 0', () => {
    const onPaginate = jest.fn(() => Promise.resolve());
    shallow(<LocationPagination history={history} location={location} onPaginate={onPaginate} />);
    expect(onPaginate.mock.calls.length).toBe(1);
  });

  it('should not calls props.onPaginate in componentWillMount when URL path is valid and props.feedCount !== 0', () => {
    const onPaginate = jest.fn(() => Promise.resolve());
    shallow(
      <LocationPagination
        history={history}
        feedCount={10}
        location={location}
        onPaginate={onPaginate}
      />,
    );
    expect(onPaginate.mock.calls.length).toBe(0);
  });

  it('should not calls props.onPaginate when location.pathname === \'/\' in componentWillMount', () => {
    const onPaginate = jest.fn();
    shallow(
      <LocationPagination
        history={history}
        location={{ pathname: '/' }}
        onPaginate={onPaginate}
      />,
    );
    expect(onPaginate.mock.calls.length).toBe(0);
  });

  it('should calls props.onPaginate when props.location.pathname is changed in componentWillReceiveProps', () => {
    const onPaginate = jest.fn(() => Promise.resolve());
    const wrapper = shallow(
      <LocationPagination
        history={history}
        location={location}
        onPaginate={onPaginate}
      />,
    );
    expect(onPaginate.mock.calls.length).toBe(1);
    wrapper.setProps({ location: { pathname: '/news/1' } });
    expect(onPaginate.mock.calls.length).toBe(1);
    wrapper.setProps({ location: { pathname: '/jobs/1' } });
    expect(onPaginate.mock.calls.length).toBe(2);
  });

  it('should calls props.history.push when simulate <Pagination /> onPaginate', () => {
    const page = 2;
    const push = jest.fn();
    const historyMock = { push };
    const wrapper = shallow(<LocationPagination history={historyMock} location={location} />);
    wrapper.find(Pagination).simulate('paginate', page);
    expect(push.mock.calls.length).toBe(1);
    expect(push.mock.calls[0]).toEqual([`/news/${page}`]);
  });
});
