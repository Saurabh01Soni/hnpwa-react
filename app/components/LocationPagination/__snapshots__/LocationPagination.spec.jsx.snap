// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationPagination /> should match snapshot when render default 1`] = `
<Pagination
  curentPage={1}
  currentPage={1}
  maxPage={1}
  onPaginate={[Function]}
/>
`;
